Increase Thrift max message size defaults

ctubbsii opened the pull request:
* Deprecate max message properties in favor of a replacement
* Simply remove the experimental ones (but they could be restored and
  deprecated like the non-experimental ones, if that's desired)
* Ensure new property defaults to `Integer.MAX_VALUE`
* Handle detecting whether a user set the new property and fall back on
  the old properties if a user customized them already
* Set the default value for the maxMessageSize to `Integer.MAX_VALUE` in
  `TServerUtils`. However, that case should only apply when the property
  passed to the method is `null`, which it never should be. A future
  improvement could validate that the property is non-null instead.
* Ensure all the max message properties that remain are added to the
  fixed properties, since they all require a restart to take effect
* Slightly simplify related code in `TabletServer` to avoid passing
  `AccumuloConfiguration` when it's already available via a class method

This fixes #3094
This fixes #3739
This fixes #3791

dlmarion commented:
We might not want to use `Integer.MAX_VALUE` exactly. I know that the limit for `byte[]` is less than that, and I think the exact number might be dependent on things out of our control. We might want to use `Integer.MAX_VALUE - 100` or something.

ctubbsii commented:
> We might not want to use `Integer.MAX_VALUE` exactly. I know that the limit for `byte[]` is less than that, and I think the exact number might be dependent on things out of our control. We might want to use `Integer.MAX_VALUE - 100` or something.

The max message size is just a count of the number of bytes read from the stream; it doesn't correspond to any particular byte array size. This is related to the max frame size in the TFramedTransport, though, and that may be related to a byte array, if the message is read fully. I'd have to look into the implementation more closely, though. In any case, we use `Integer.MAX_VALUE` elsewhere for client side, and it really doesn't matter if the value here exceeds the maximum array index limit. The point is to make it so impractically large that it won't trigger unnecessary errors for users. If they hit an error because they max out a byte array, or because they hit this limit, it's already kind of catastrophic, and it really doesn't matter which error specifically caused the catastrophe (actually, it'd probably better if it was a JVM one, because that would be clearly distinguishable from the EOFException that Thrift causes, which looks the same as a transient network error; at least the JVM complaining about a byte array being too big would stand out as a distinct problem).

In any case, this is still WIP. I need to do more here, but I quickly put something up to demo the end-to-end process of cloning a repo, making a change, and creating a pull request, for the benefit of a newer developer.

ctubbsii commented:
@dlmarion This is ready for review. I updated the PR description at the top. In particular, I want to draw your attention to the decision that should be made to either drop the experimental properties, or have them go through deprecation as well. I dropped them in this PR, but I can restore them if anybody feels like we need to keep them, deprecated, for the remainder of the 2.1 lifecycle.

dlmarion commented:
> @dlmarion This is ready for review. I updated the PR description at the top. In particular, I want to draw your attention to the decision that should be made to either drop the experimental properties, or have them go through deprecation as well. I dropped them in this PR, but I can restore them if anybody feels like we need to keep them, deprecated, for the remainder of the 2.1 lifecycle.

I have no issue with removing them as long as it doesn't break any user that may have them in their configuration.

ctubbsii commented on server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java:
```diff
@@ -153,7 +153,7 @@ public static ServerAddress startServer(ServerContext context, String hostname,
     }
 
-    long maxMessageSize = 10_000_000;
+    long maxMessageSize = Integer.MAX_VALUE;
```
I commented on the issue. It's not possible in the configuration. It is possible we can hard-code some negative value somewhere, but that would be a bug in the code, not something a user could do.

ctubbsii commented:
> I have no issue with removing them as long as it doesn't break any user that may have them in their configuration.

In my mind, users have already accepted the risks by choosing to engage with features marked "experimental". We've already signaled that a break might occur if they grow to depend on an experimental feature. That's the purpose of signaling with the experimental flag. However, I think the risks of breakage here are very small. That said, it really depends on what "break" means for a particular user. If it means that they were expecting message sizes to be limited by a strict upper bound, this change will "break" them in that sense (because it changes the default in most cases to a much larger value than anything the user might have set in a property that is now ignored). Similarly, if the user is monitoring for errors or warnings about unrecognized properties in the system configuration, or has scripts to set the property, which is now no longer a valid system configuration, then that could be considered a break because they won't be able to set that property in ZK anymore. But I think these are stretching what it means to be "broken". In the typical case, this will relieve users from the burden of setting any value at all here, that they might have previously had to do to increase it from the default.
